libgalois/include/katana/EntityTypeManager.h

```diff
@@ -0,0 +1,227 @@
+        entity_type_id_to_atomic_entity_type_ids_(
+            std::move(entity_type_id_to_atomic_entity_type_ids)) {
+    for (auto& type_id_name_pair : atomic_entity_type_id_to_type_name_) {
+      atomic_type_name_to_entity_type_id_.emplace(
+          std::make_pair(type_id_name_pair.second, type_id_name_pair.first));
+    }
+
+    size_t num_entity_types = entity_type_id_to_atomic_entity_type_ids_.size();
+    atomic_entity_type_id_to_entity_type_ids_.resize(num_entity_types);
+    for (size_t i = 0, ni = num_entity_types; i < ni; ++i) {
+      for (size_t j = 0, nj = num_entity_types; j < nj; ++j) {
+        if (entity_type_id_to_atomic_entity_type_ids_[i].test(j)) {
+          atomic_entity_type_id_to_entity_type_ids_[j].set(i);
+        }
+      }
+    }
+  }
+
+  EntityTypeManager(
+      EntityTypeIDToAtomicTypeNameMap&& atomic_entity_type_id_to_type_name,
+      AtomicTypeNameToEntityTypeIDMap&& atomic_type_name_to_entity_type_id,
+      EntityTypeIDToSetOfEntityTypeIDsMap&&
+          entity_type_id_to_atomic_entity_type_ids,
+      EntityTypeIDToSetOfEntityTypeIDsMap&&
+          atomic_entity_type_id_to_entity_type_ids)
+      : atomic_entity_type_id_to_type_name_(
+            std::move(atomic_entity_type_id_to_type_name)),
+        atomic_type_name_to_entity_type_id_(
+            std::move(atomic_type_name_to_entity_type_id)),
+        entity_type_id_to_atomic_entity_type_ids_(
+            std::move(entity_type_id_to_atomic_entity_type_ids)),
+        atomic_entity_type_id_to_entity_type_ids_(
+            std::move(atomic_entity_type_id_to_entity_type_ids)) {}
+
+  void Reset() {
+    atomic_entity_type_id_to_type_name_.clear();
+    atomic_type_name_to_entity_type_id_.clear();
+    entity_type_id_to_atomic_entity_type_ids_.clear();
+    atomic_entity_type_id_to_entity_type_ids_.clear();
+    Init();
+  }
+
+  /// adds a new entity type for the atomic type with name \p name
+  /// \returns the EntityTypeID for the new type
+  EntityTypeID AddAtomicEntityType(const std::string& name) {
+    KATANA_LOG_DEBUG_ASSERT(!HasAtomicType(name));
+
+    // TODO(roshan) limit the new entity type id to the max for 8 bits
+    // and return an error
+    EntityTypeID new_entity_type_id = GetNumEntityTypes();
+    atomic_entity_type_id_to_type_name_.emplace(
+        std::make_pair(new_entity_type_id, name));
+    atomic_type_name_to_entity_type_id_.emplace(
+        std::make_pair(name, new_entity_type_id));
+
+    SetOfEntityTypeIDs entity_type_ids;
+    entity_type_ids.set(new_entity_type_id);
+    entity_type_id_to_atomic_entity_type_ids_.emplace_back(entity_type_ids);
+    atomic_entity_type_id_to_entity_type_ids_.emplace_back(entity_type_ids);
+
+    return new_entity_type_id;
+  }
+
+  /// adds a new entity type for the set of atomic types with names \p names
+  /// \returns the EntityTypeID for the new type
+  EntityTypeID AddNonAtomicEntityType(const std::vector<std::string>& names) {
```

Comment:
@witchel The above two functions provide ways to add more types

Reply:
I think this function should take EntityTypeIDs instead of names.

Reply:
I agree except that it limits the interface some. Currently, this function creates any necessary atomic types that don't exist. If we passed `EntityTypeID`s it wouldn't be able to do that. Then again, maybe it would be a good thing to explicitly force users to do that as two separate actions.

Reply:
Agree with Arthur. It is "surprising" that the function will create necessary atomic types.